[premake5] Add new port

thomaskrause opened the pull request:
**Describe the pull request**
This adds the new package premake5, which is a build file generator similiar to cmake but based on simple Lua scripts

- What does your PR fix? 
 No packaging request was added as an issue as far as I know.

- Which triplets are supported/not supported? Have you updated the CI baseline?
The build configurations included in the released tar-balls do not support arm/arm64 and there are only build files for Windows, MacOS and Linux. I used the `Supports` field in the control file to define the allowed triples.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
I choose to use GNU Make on Linux/Mac and msbuild on Windows because that is the native build system used by premake5 and it would have been difficult to map this to CMake. Also, I'm not using the Git repository directly, but the released tar-balls because only these include the pre-generated project files (additional bootstrapping would be needed when using the Git checkout directly).
I'm not sure if I handled the version number correctly. The original authors are using the "alpha 15" suffix (where the number will increment) and use spaces. I decided to use the Git tag `v5.0.0-alpha15` as a base instead, which removes the spaces.

thomaskrause committed:
Initial port of premake5, only configured for and tested on Linux

thomaskrause committed:
Configure for MacOS and Windows

thomaskrause committed:
Fix Windows build configuration

thomaskrause committed:
Use tag version for less spaces

thomaskrause committed:
Allow arm configurations to fail

thomaskrause committed:
Use the control file to define which platforms are supported (arm/arm64 is not)

thomaskrause committed:
Add new line to control file

Voskrese commented:
now this is unnecessary

` include(vcpkg_common_functions)`

thomaskrause committed:
Removed unnecessary inclusion of common functions

thomaskrause commented:
> now this is unnecessary
> 
> ` include(vcpkg_common_functions)`

Thanks for the hint, I don't know if i introduced this line during debugging to fix another error (that was probably unrelated) but its removed now. x64_osx pipeline is failing because the Git checkout failed. I hope this is a temporary failure, compiling premake5 should also work on OSX (and did work in earlier commits when the git checkout succeeded)

Voskrese commented:
this is a Ci error, wait for it to be fixed

JackBoosY commented on ports/premake5/portfile.cmake:
```diff
@@ -0,0 +1,46 @@
+if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
+        set(BUILD_CONFIG "debug")
+endif()
+set(BUILD_CONFIG "release")
```
This line should be added to `else()`

thomaskrause replied:
Fixed in 0f0cc20b1dde6dd5e9820a099315466accdce16b

JackBoosY commented on ports/premake5/portfile.cmake:
```diff
@@ -0,0 +1,46 @@
+set(VCPKG_POLICY_EMPTY_INCLUDE_FOLDER enabled)
```
Is there no header file available for this port?

thomaskrause replied:
Yes, this port produces an executable ("premake5") as main result. I can add a comment to the portfile with an explanation if you want.

JackBoosY replied:
@thomaskrause See #10505.

thomaskrause replied:
Ok, I will look into that example and see if I can adapt this PR.

JackBoosY commented on ports/premake5/portfile.cmake:
```diff
@@ -0,0 +1,46 @@
+        PROJECT_SUBPATH build/vs2015/Premake5.vcxproj
+    )
+elseif(VCPKG_TARGET_IS_OSX)
+    find_program(MAKE make REQUIRED)
```
We already have functions `vcpkg_*_make`.
Please see documentation:
https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_configure_make.md
https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_build_make.md
https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_install_make.md

Example: [tcl](https://github.com/Microsoft/vcpkg/blob/master/ports/tcl/portfile.cmake)

thomaskrause replied:
I could not get `vcpkg_build_make` to work because it needs the `vcpkg_configure_make` script to run first. These makefiles are not generated with autoconf and there are nor configure scripts. I tried to use it with `SKIP_CONFIGURE` like in this example:
```cmake
    vcpkg_configure_make(
        SKIP_CONFIGURE
        PROJECT_SUBPATH ${SOURCE_PATH}/build/gmake2.unix/
    )
    vcpkg_build_make(TARGET "Premake5")
```
Unfortunately i consistently get an error when trying this:
```
-- Using generator make
CMake Error at scripts/cmake/vcpkg_configure_make.cmake:246 (file):
  file COPY cannot find "/dev/disk/by-partlabel/EFI/x20System/x20Partition".
Call Stack (most recent call first):
  ports/premake5/portfile.cmake:34 (vcpkg_configure_make)
  scripts/ports.cmake:90 (include)
```

thomaskrause replied:
I could resolve the issue with `vcpkg_build_make` and I am using it now on Linux and OSX builds (fd4a673476c32734b42b31e5dc120fd1249457ad)

DailyShana commented:
premake5 is not a library.
Is it acceptable to add non-libraries to vcpkg?

JackBoosY commented:
@DailyShana If premake5 is a build tool, I prefer to add it to the tool list rather than a port.

thomaskrause committed:
Use else block to define "release" as default

thomaskrause commented:
> @DailyShana If premake5 is a build tool, I prefer to add it to the tool list rather than a port.

Yes, premake5 is indeed a build tool

thomaskrause committed:
Use vcpkg_build_make on Linux and Mac, but only compile in release mode.

Its difficult to get the right binary if there are both debug and release builds and `vcpkg_build_make` will compile out of place. Also, the premake makefiles do not support an "install" target.
Instead of reconstructing the logic in `vcpkg_build_make`, we just disable debug builds.
Since premake5 is an executable, this should not be an issue.

thomaskrause committed:
Fix missing space

JackBoosY commented:
@thomaskrause Please ping me if this PR is ready for review.

Thanks.
